File: benchmark_client/scripts/cpp/Invert Matrices.cpp
#include <chrono>
#include <cmath>
#include <cstddef>
#include <fstream>
#include <iostream>
#include <stdexcept>
#include <string>
#include <vector>

static void read_matrix(const char* filename, int& n, std::vector<float>& data) {
    std::ifstream file(filename, std::ios::binary);
    if (!file.is_open()) {
        std::cerr << "Error: File not found\n";
        exit(1);
    }

    file.read(reinterpret_cast<char*>(&n), sizeof(int));
    if (!file || n <= 0) {
        throw std::runtime_error(std::string("Failed to read matrix size from ") + filename);
    }

    const std::size_t total = static_cast<std::size_t>(n) * static_cast<std::size_t>(n);
    data.resize(total);
    file.read(reinterpret_cast<char*>(data.data()), static_cast<std::streamsize>(total * sizeof(float)));
    if (!file) {
        throw std::runtime_error(std::string("Failed to read matrix data from ") + filename);
    }
}

static void write_matrix(const char* filename, int n, const std::vector<float>& data) {
    std::ofstream file(filename, std::ios::binary);
    if (!file) {
        throw std::runtime_error(std::string("Failed to open ") + filename);
    }

    const std::size_t total = static_cast<std::size_t>(n) * static_cast<std::size_t>(n);
    if (data.size() != total) {
        throw std::runtime_error("Invalid output matrix size.");
    }

    file.write(reinterpret_cast<const char*>(&n), sizeof(int));
    file.write(reinterpret_cast<const char*>(data.data()), static_cast<std::streamsize>(total * sizeof(float)));
}

static bool invert_gauss_jordan(const std::vector<float>& input, int n, std::vector<float>& inverse) {
    const std::size_t n_sz = static_cast<std::size_t>(n);
    const std::size_t cols = n_sz * 2;
    std::vector<float> aug(n_sz * cols, 0.0f);

    for (int i = 0; i < n; ++i) {
        const std::size_t i_row = static_cast<std::size_t>(i) * cols;
        const std::size_t in_row = static_cast<std::size_t>(i) * n_sz;
        for (int j = 0; j < n; ++j) {
            aug[i_row + static_cast<std::size_t>(j)] = input[in_row + static_cast<std::size_t>(j)];
        }
        aug[i_row + n_sz + static_cast<std::size_t>(i)] = 1.0f;
    }

    for (int i = 0; i < n; ++i) {
        int pivot_row = i;
        float pivot_abs = std::fabs(aug[static_cast<std::size_t>(i) * cols + static_cast<std::size_t>(i)]);
        for (int r = i + 1; r < n; ++r) {
            const float candidate =
                std::fabs(aug[static_cast<std::size_t>(r) * cols + static_cast<std::size_t>(i)]);
            if (candidate > pivot_abs) {
                pivot_abs = candidate;
                pivot_row = r;
            }
        }

        if (pivot_abs <= 1e-12f) {
            return false;
        }

        if (pivot_row != i) {
            const std::size_t i_off = static_cast<std::size_t>(i) * cols;
            const std::size_t p_off = static_cast<std::size_t>(pivot_row) * cols;
            for (std::size_t c = 0; c < cols; ++c) {
                const float tmp = aug[i_off + c];
                aug[i_off + c] = aug[p_off + c];
                aug[p_off + c] = tmp;
            }
        }

        const std::size_t i_off = static_cast<std::size_t>(i) * cols;
        const float pivot = aug[i_off + static_cast<std::size_t>(i)];
        for (std::size_t c = 0; c < cols; ++c) {
            aug[i_off + c] /= pivot;
        }

        for (int r = 0; r < n; ++r) {
            if (r == i) {
                continue;
            }
            const std::size_t r_off = static_cast<std::size_t>(r) * cols;
            const float factor = aug[r_off + static_cast<std::size_t>(i)];
            if (factor == 0.0f) {
                continue;
            }
            for (std::size_t c = 0; c < cols; ++c) {
                aug[r_off + c] -= factor * aug[i_off + c];
            }
        }
    }

    inverse.resize(n_sz * n_sz);
    for (int i = 0; i < n; ++i) {
        const std::size_t i_off = static_cast<std::size_t>(i) * cols;
        const std::size_t out_off = static_cast<std::size_t>(i) * n_sz;
        for (int j = 0; j < n; ++j) {
            inverse[out_off + static_cast<std::size_t>(j)] = aug[i_off + n_sz + static_cast<std::size_t>(j)];
        }
    }

    return true;
}

int main() {
    try {
        int n = 0;
        std::vector<float> t;
        read_matrix("T.dat", n, t);

        std::vector<float> inv;
        const auto start = std::chrono::high_resolution_clock::now();
        const bool ok = invert_gauss_jordan(t, n, inv);
        const auto end = std::chrono::high_resolution_clock::now();
        if (!ok) {
            return 1;
        }

        const double elapsed_ms = std::chrono::duration<double, std::milli>(end - start).count();
        write_matrix("M3.dat", n, inv);
        std::cout << "{\"execution_time_ms\": " << elapsed_ms << "}\n";
        return 0;
    } catch (const std::exception&) {
        return 1;
    }
}

File: benchmark_client/scripts/go/Invert Matrices.go
package main

import (
	"encoding/binary"
	"fmt"
	"math"
	"os"
	"time"
)

func readMatrix(path string) (int, []float32, error) {
	file, err := os.Open(path)
	if err != nil {
		return 0, nil, err
	}
	defer file.Close()

	var n int32
	if err := binary.Read(file, binary.LittleEndian, &n); err != nil {
		return 0, nil, err
	}
	if n <= 0 {
		return 0, nil, fmt.Errorf("invalid matrix size")
	}

	total := int(n) * int(n)
	data := make([]float32, total)
	if err := binary.Read(file, binary.LittleEndian, data); err != nil {
		return 0, nil, err
	}

	return int(n), data, nil
}

func writeMatrix(path string, n int, data []float32) error {
	file, err := os.Create(path)
	if err != nil {
		return err
	}
	defer file.Close()

	if len(data) != n*n {
		return fmt.Errorf("invalid output matrix size")
	}

	n32 := int32(n)
	if err := binary.Write(file, binary.LittleEndian, n32); err != nil {
		return err
	}
	if err := binary.Write(file, binary.LittleEndian, data); err != nil {
		return err
	}
	return nil
}

func invertGaussJordan(input []float32, n int) ([]float32, bool) {
	cols := 2 * n
	aug := make([]float32, n*cols)

	for i := 0; i < n; i++ {
		iOff := i * cols
		inOff := i * n
		for j := 0; j < n; j++ {
			aug[iOff+j] = input[inOff+j]
		}
		aug[iOff+n+i] = 1.0
	}

	for i := 0; i < n; i++ {
		pivotRow := i
		pivotAbs := float32(math.Abs(float64(aug[i*cols+i])))
		for r := i + 1; r < n; r++ {
			v := float32(math.Abs(float64(aug[r*cols+i])))
			if v > pivotAbs {
				pivotAbs = v
				pivotRow = r
			}
		}
		if pivotAbs <= 1e-12 {
			return nil, false
		}

		if pivotRow != i {
			iOff := i * cols
			pOff := pivotRow * cols
			for c := 0; c < cols; c++ {
				aug[iOff+c], aug[pOff+c] = aug[pOff+c], aug[iOff+c]
			}
		}

		iOff := i * cols
		pivot := aug[iOff+i]
		for c := 0; c < cols; c++ {
			aug[iOff+c] /= pivot
		}

		for r := 0; r < n; r++ {
			if r == i {
				continue
			}
			rOff := r * cols
			factor := aug[rOff+i]
			if factor == 0 {
				continue
			}
			for c := 0; c < cols; c++ {
				aug[rOff+c] -= factor * aug[iOff+c]
			}
		}
	}

	inv := make([]float32, n*n)
	for i := 0; i < n; i++ {
		iOff := i * cols
		outOff := i * n
		for j := 0; j < n; j++ {
			inv[outOff+j] = aug[iOff+n+j]
		}
	}

	return inv, true
}

func main() {
	n, t, err := readMatrix("T.dat")
	if err != nil {
		os.Exit(1)
	}

	start := time.Now()
	inv, ok := invertGaussJordan(t, n)
	elapsedMs := float64(time.Since(start).Nanoseconds()) / 1e6
	if !ok {
		os.Exit(1)
	}

	if err := writeMatrix("M3.dat", n, inv); err != nil {
		os.Exit(1)
	}
	fmt.Printf("{\"execution_time_ms\": %v}\n", elapsedMs)
}
